fix(trace): propagate trace from pubsub msg

kutkaite opened the pull request:
this ensures that tracing from the pubsub message is properly propagated before we create a context with log tracing. Without this fix, the cloudrequestlog middleware logs incorrectly and incorrect traceId is attached to the slog field with TraceHook.

There’s no reliable way to determine whether the request originates from pubsub without attempting to decode it. I could make this configurable to avoid propagating the trace when it's not wanted.

Let’s discuss if you have any better ideas for addressing this!

kutkaite commented:
There are maybe headers included https://cloud.google.com/pubsub/docs/payload-unwrapping#unwrapped_message_with_write_metadata_enabled I'll explore a bit on that and see if we could narrow down the requests that need to be decoded

radhus commented on cloudotel/tracemiddleware.go:
```diff
@@ -97,3 +105,36 @@ func (i *TraceMiddleware) withLogTracing(ctx context.Context, spanCtx trace.Span
+	}
+	bodyBuffer, payload, err := readAndDecodeRequestBody(r)
+	if err != nil {
+		return ctx
```
Won't we have emptied the `r.Body` reader in this branch? I.e. should we make sure to use `TeeReader` before calling `readAndDecodeRequestBody`?

kutkaite replied:
yea this is wrong, good catch! Will fix

radhus reviewed (commented):
> There’s no reliable way to determine whether the request originates from pubsub without attempting to decode it. I could make this configurable to avoid propagating the trace when it's not wanted.
>
> Let’s discuss if you have any better ideas for addressing this!

I think the setup in your PR is a fair opinionated default! 👍 Let's add an option if someone really wants to opt-out.

I'm just worried a little bit about the body reader handling, would be nice with some tests that verifies incoming mocked requests with and without pubsub messages, and verifies that the HTTP handler can read the body in both cases.

radhus commented on cloudotel/tracemiddleware.go:
```diff
@@ -97,3 +105,36 @@ func (i *TraceMiddleware) withLogTracing(ctx context.Context, spanCtx trace.Span
+	tr := io.TeeReader(r.Body, &bodyBuffer)
+	var payload cloudpubsub.Payload
+	if err := json.NewDecoder(tr).Decode(&payload); err != nil {
+		return bodyBuffer, payload, err
```
Also in relation to comment above - I'm not sure here, but I don't think `json.Decode()` guarantees it will read _all_ bytes of the provided reader `tr`? Since `bodyBuffer` will only contain the bytes read from `tr`, the consumer of `bodyBuffer.Read()` might not get all of the provided HTTP body 🧐

kutkaite replied:
fair enough, I'll test!

kutkaite replied:
I looked into it and `json.Decode()` reads the whole value into a buffer before reading the json, so if that's successful then we consume the stream, if this part fails then the whole read fails. It also makes sure to completely read the json object even if an error happens, so in case of a malformed JSON we shouldn't lose any bytes. 
I thought to do something like:
```
	pubsubPayload, err := tryUnmarshalAsPubsubPayload(tr)
	if err != nil {
		if len(bodyBuffer.Bytes()) != 0 {
			r.Body = io.NopCloser(bytes.NewReader(bodyBuffer.Bytes()))
		}
		return ctx
	}
```

But, if we don't want to risk it, I can just use `io.ReadAll(r.Body)` instead of `io.TeeReader` and reassign r.Body to the buffer right away. Not as memory efficient, but this whole thing is already not efficient 😄 What do you think?

kutkaite replied:
I changed it to `io.ReadAll(r.Body)` for now, but can change back

kutkaite commented:
Added a couple of tests, changed from `io.TeeReader` to `io.ReadAll(r.Body)` and added an option to disable this in the runtime config ✅ 
there are no headers that I could use instead. 

Tested this on a Cloud Run service with mixed GRPC and HTTP traffic. All seems to work as expected with various configurations (enabled and disabled `EnableOpenTelemetryTracing` on a PubSub client, and non PubSub HTTP requests). I also tested manually disabling PubSub tracing in the runtime config.

radhus commented on cloudotel/tracemiddleware_test.go:
Nice!

radhus reviewed (commented):
Code looks good given it's only used for pubsub endpoints, so I would approve with a default disable 👍

kutkaite commented on cloudruntime/config.go:
```diff
@@ -43,6 +43,8 @@ type Config struct {
 	// ServiceVersion is the version of the service.
 	ServiceVersion string `env:"SERVICE_VERSION"`
+	// EnablePubsubTracing, enabled by default, reads trace parent from Pub/Sub message attributes.
+	EnablePubsubTracing bool `env:"ENABLE_PUBSUB_TRACING"`
```
makes sense!

radhus reviewed (approved):
Awesome work! ⭐
